Add basic AiMod-like features

Naxesss opened the pull request:
Makes progress towards resolving https://github.com/ppy/osu/issues/12091.

This adds
- A "verify" tab
- Check and issue classes
- 2 realistic checks
    - Background presence (ruleset-agnostic)
    - Offscreen hitobjects (osu-specific)

Am thinking of adding tests to those checks too, but I want to gather thoughts on the overall design first.

### Verify tab
Looks like this at the moment:

![image](https://user-images.githubusercontent.com/30292137/114270655-db558780-9a0d-11eb-940a-b815e7ec2a49.png)

Am thinking we could further develop the UI as a secondary pass like @peppy [mentioned here](https://github.com/ppy/osu/issues/12091#issuecomment-803618170). It's still missing things like check parents, truncation, show more/less of same issues, settings, etc. meaning it'll get pretty messy once we add more checks. At the moment it's basically just using the same UI as the timing tab.

### Check and issue classes
Are structured like this:

![image](https://user-images.githubusercontent.com/30292137/114274417-9554ef80-9a1e-11eb-85ab-dd79601e6872.png)

I've added a method in `Ruleset` giving the respective `Checker` (default null), similarly to `HitObjectComposer`, so that should be properly decoupled I think:
```cs
public virtual HitObjectComposer CreateHitObjectComposer() => null;

public virtual Checker CreateChecker() => null;
```

### The 2 checks added so far
Can be tested by simply entering Edit without setting anything, placing a slider that goes offscreen, then looking at the verify tab.

---

Side-note: I noticed that pressing ctrl automatically changes tab to verify and I'm not sure why...

Naxesss committed:
Add verify tab

Currently empty, but works.

Naxesss committed:
Add check/issue classes

Naxesss committed:
Add ruleset-specific checker

Naxesss committed:
Add example checks

Naxesss committed:
Update UI with new components

Naxesss committed:
Remove visibility settings

Can look into this later, not really important for a first iteration.

Naxesss committed:
Move components to more appropriate spot

Naxesss committed:
Fix comment; mode -> ruleset

Naxesss committed:
Replace checks with realistic ones

Naxesss committed:
Add issue selection

This mainly helps with keeping track of which issue was clicked, since doing so switches tab.

Naxesss committed:
Remove todo

Doesn't really matter in the end, as only one checker will run at a time in this case.

Naxesss committed:
Show table even if no issues

Naxesss committed:
Improve table formatting

Naxesss committed:
Merge `Check` and `BeatmapCheck`

We're probably not going to need GeneralChecks or BeatmapsetChecks.

The verify tab is only available to a single difficulty at a time, and we already have access to the rest of the set through `IBeatmap`.

Naxesss committed:
Remove old todo

Naxesss committed:
Fix field name and accessibility

peppy commented:
Without checking the code, the UI looks fine for a first pass, nice stuff.

Only thing I'd initially comment on is that I would hope a "refresh" button would not be required, and it was automatically refreshed (or just inherently always up-to-date).

Naxesss commented:
I was thinking about something like that as well, but couldn't find a way to subscribe to the map being changed (including hitobjects / timing / metadata / files / etc). May need some pointers there.

peppy commented:
You'd want to look at the `EditorChangeHandler` (which I believe is now implemented via `EditorBeatmap`, which is considered the best way to interact with it). As you mention, it only handles `HitObject` related changes right now, but this will be expanded as the diffing algorithm is improved upon)

Naxesss commented:
Probably best to hold off on that for now then. `EditorChangeHandler` doesn't seem to actually catch all changes for hitobjects either, so that could easily get confusing if the user expects it to refresh the verify tab and it doesn't.

Here I set the button alpha to full when `changeHandler.OnStateUpdate` happens, and I expected it to become opaque every time, whereas it only did sometimes (probably because it considers 2 separate movements as within the same "state" or something):

https://user-images.githubusercontent.com/30292137/114299288-b703b480-9aba-11eb-94b1-09c7cb42c344.mp4

Either way, I'm thinking this can probably come as a tertiary pass once things are more fleshed out - that way we can test its performance with more checks too, which is my main concern with this approach.

peppy commented:
That's very weird. Might be something to do with the screen switching logic.

peppy commented:
Just for reference, we are tracking the change handling not handling everything it should at #10314 (was trying to find this earlier when commenting but couldn't).

peppy committed:
Merge branch 'master' into verify-tab

peppy committed:
Checker -> BeatmapVerifier

peppy committed:
Wrap xmldoc less and make a few fixes

peppy committed:
Colour constants to private

peppy committed:
Fix formatting of colour getter

peppy committed:
Move enums out of nesting

peppy committed:
Change some methods to get properties

peppy committed:
Tidy some remaining code

peppy committed:
Change Metadata into a get property

peppy commented on osu.Game/Rulesets/Edit/Checks/CheckBackground.cs:
```diff
@@ -0,0 +1,57 @@
+            templateDoesNotExist
+        };
+
+        private readonly IssueTemplate templateNoneSet = new IssueTemplate
+        (
+            type: IssueType.Problem,
+            unformattedMessage: "No background has been set."
+        );
+
+        private readonly IssueTemplate templateDoesNotExist = new IssueTemplate
+        (
+            type: IssueType.Problem,
+            unformattedMessage: "The background file \"{0}\" is does not exist."
+        );
```
These should be constructed on use. Probably a good case to make nested classes for each of these. `PossibleTemplates` would just construct a copy of each class.

Naxesss replied:
Changed to nested classes, where `PossibleTemplates` receives copies of these constructed in the constructor of the check. This way the check can reuse the template to produce multiple issues, rather than creating a new instance for each.

This approach also helps with the other two suggestions.

peppy commented on osu.Game/Rulesets/Edit/Checks/Components/Check.cs:
```diff
@@ -0,0 +1,33 @@
+
+        protected Check()
+        {
+            foreach (var template in PossibleTemplates)
+                template.Origin = this;
```
As discussed, remove this and move the `Check` reference to the `Issue` itself. I'd also call it `Check`, not `Origin`.

Naxesss replied:
I've removed this part of the code, but decided to keep `Check` in `IssueTemplate`, as the it's accessible from `Issue` (which I've made a bit more convenient with a property):
```cs
public ICheck Check => Template.Check;
```
This way we don't need to explicitly provide the check for each issue constructed from the template, and instead only provide it to the template `ctor`.

peppy commented on osu.Game/Rulesets/Edit/Checks/Components/IssueType.cs:
```diff
@@ -0,0 +1,25 @@
+    /// </summary>
+    public enum IssueType
+    {
+        /// <summary> A must-fix in the vast majority of cases. </summary>
+        Problem = 3,
+
+        /// <summary> A possible mistake. Often requires critical thinking.  </summary>
+        Warning = 2,
+
+        // TODO: Try/catch all checks run and return error templates if exceptions occur.
+        /// <summary> An error occurred and a complete check could not be made. </summary>
+        Error = 1,
+
+        // TODO: Negligible issues should be hidden by default.
+        /// <summary> A possible mistake so minor/unlikely that it can often be safely ignored. </summary>
+        Negligible = 0,
```
Not sure about the reverse ordering here. Is that done for a reason? Probably best to order ascending to follow standard (and apply sorting at some other level if it's required)

peppy commented:
I've applied some code quality fixes and as discussed, made a few proposals for improving the structure.

I'll also look into the tab selector selecting this when hitting `Ctrl` as you discovered.

peppy commented:
In addition, I'm not sure about the non-typed passing of arguments as you have it right now. Optimally, they would be typed with derived implementations of `IssueTemplate` (and passed in via the `ctor`).

peppy committed:
Add binding for verify mode (and move enum entry to end)

peppy commented:
The `Control` key activating the screen is due to adding a new enum value in the middle of `GlobalActions`. Right now these are serialised as `int` so by adding this you've offset all the stored key bindings in the database. I've moved it to the end.

Naxesss committed:
Move check origin from `IssueTemplate` to `Issue`

As a result we can also make check an interface, and need to provide the check itself when constructing an issue.

Naxesss committed:
Rename `CreateChecker` -> `CreateBeatmapVerifier`

Naxesss committed:
Rework template usage

Includes moving the origin check back to templates, constructing nested template classes in each check, and making parameterized template usage.

Naxesss committed:
Fix `Template.Origin` -> `Check`

Naxesss committed:
Reverse `IssueType` ordering

Reversed both in the enum and where it's displayed, so ends up the same in the end.

peppy commented on osu.Game/Rulesets/Edit/Checks/Components/IssueTemplate.cs:
```diff
@@ -0,0 +1,74 @@
+        /// <summary>
+        /// The check that this template originates from.
+        /// </summary>
+        public ICheck Check;
```
this can be `readonly` right?

Naxesss replied:
yes, will fix

Naxesss committed:
Rename `Check` -> `ICheck`

Naxesss committed:
Make `IssueTemplate.Check` readonly

Naxesss commented:
> In addition, I'm not sure about the non-typed passing of arguments as you have it right now. Optimally, they would be typed with derived implementations of IssueTemplate (and passed in via the ctor).

I couldn't have the args be passed via the ctor, as that would prevent their construction for `PossibleTemplates`, but I've added `Create` methods in the nested classes as parameterized ways to produce their issues. Let me know how that looks.

peppy committed:
Move nested classes to bottom of file

peppy committed:
Remove redundant parameter naming

peppy commented:
Honestly, I'd just do it like this and avoid all the locals and array storage:

```csharp

    public class CheckBackground : ICheck
    {
        public CheckMetadata Metadata { get; } = new CheckMetadata(CheckCategory.Resources, "Missing background");

        public IEnumerable<IssueTemplate> PossibleTemplates => new IssueTemplate[]
        {
            new IssueTemplateNoneSet(this),
            new IssueTemplateDoesNotExist(this)
        };

        public IEnumerable<Issue> Run(IBeatmap beatmap)
        {
            if (beatmap.Metadata.BackgroundFile == null)
            {
                yield return new IssueTemplateNoneSet(this).Create();

                yield break;
            }

            // If the background is set, also make sure it still exists.

            var set = beatmap.BeatmapInfo.BeatmapSet;
            var file = set.Files.FirstOrDefault(f => f.Filename == beatmap.Metadata.BackgroundFile);

            if (file != null)
                yield break;

            yield return new IssueTemplateDoesNotExist(this).Create(beatmap.Metadata.BackgroundFile);
        }

```

The call to `PossibleTemplates` should be so rare that this is not going to be a huge issue. More worried about keeping the code as concise as possible for now. Still not sure about the `this` passing yet, will need to think about that a touch longer.

Naxesss committed:
Remove check ctors and locals

Naxesss committed:
Rename `checkOrigin` -> `check`

More consistent with `Issue.ctor`'s "template".

peppy commented:
I think this is feeling pretty good for a first pass implementation. Will do a final review pass on it tomorrow 👍🏻.

Naxesss committed:
Add `Equals` method to `IssueTemplate`

This will be useful in tests.

Naxesss committed:
Add `CheckBackground` tests

Naxesss committed:
Add `CheckOffscreenObjects` tests

Naxesss commented:
I've added some unit tests for each check.

It's worth noting that they rely on the ordering of `PossibleTemplates` for finding the right template. It would be great if something like this was possible:
```cs
Assert.That(issue.Template is CheckOffscreenObjects.IssueTemplateOffscreenSlider);
```
but the nested classes should probably stay private, else it'd be possible to use them outside of their intended context. There are ways to prevent the `ctor` and methods from being used outside of the enclosing class, but the approaches I've seen are likely not worth the added complexity.

Naxesss committed:
Remove redundant space

frenzibyte commented:
I believe it can be exposed for that reason, instead of forcing the possible templates to follow a certain arbitrary order and potentially regress tests from retrieving the right template.

Naxesss committed:
Use `is` class instead of `Equals` with template index

Ensures ordering of `PossibleTemplates` does not affect tests.

Naxesss commented:
Hmm that's true, risk of someone misusing it is probably lower than us accidentally changing the order of that list, have changed accordingly.

peppy commented:
Exposing is fine. If you want to limit construction you can still make the constructor `private`.

peppy committed:
Use static for playfield centre positioning

peppy commented on osu.Game.Rulesets.Osu/Edit/Checks/CheckOffscreenObjects.cs:
```diff
@@ -0,0 +1,114 @@
+    {
+        // These are close approximates to the edges of the screen
+        // in gameplay on a 4:3 aspect ratio for osu!stable.
+        private const int min_x = -67;
+        private const int min_y = -60;
+        private const int max_x = 579;
+        private const int max_y = 428;
```
Do you have a reference for how these were calculated? Probably something we want to get right from the start else it may be overlooked until it crops up in some mapping drama.

Naxesss replied:
Reference is a mix between cross-referencing with previous modding tools, my own testing using slider lengths in 1 px increments for every side of the screen, and 2 years of feedback from Mapset Verifier using these numbers (haven't had anyone report a false positive or negative for this check in a long time, and it's used by practically all BNs).

Only thing I'm unsure of is if the coords convert to lazer 1:1, but I suppose they would have to.

peppy replied:
So these are based on gameplay coordinate space, correct? (512x384 playfield / 640x480 screen area)

peppy committed:
Update asserts to use better nunit specifications

peppy commented:
I've updated your `Assert` calls to use more correct nunit specifications (generally just gives better error messages on failure, since it has more context to work with).

Naxesss commented on osu.Game.Rulesets.Osu/Edit/OsuBeatmapVerifier.cs:
```diff
@@ -0,0 +1,30 @@
+
+        public override IEnumerable<Issue> Run(IBeatmap beatmap)
+        {
+            // Also run mode-invariant checks.
+            foreach (var issue in base.Run(beatmap))
+                yield return issue;
```
I think the only situation they would want to silence ruleset-invariant checks is if their ruleset overrides parts of the general Ranking Criteria which none of the 4 base modes do, for example:
- Allowing higher resolution backgrounds than regular.
- Being more lenient with drain time requirements.
- Not requiring maps to be hitsounded.
- Allowing audio files to be >192 kbps.

So if they are allowed to deviate like that, then this would be a case of that. The majority of ruleset-invariant checks would probably be kept, though. So it would end up as something like:
```cs
foreach (var issue in base.Run(beatmap))
{
    if (!(issue.Check is CheckHitsoundVolume || issue.Check is CheckSpreadRequirements))
        yield return issue;
}
```

An alternative would be simply constructing general checks we think may want to be overridden in each ruleset-specific `BeatmapVerifier` instead, and then separating the rest one level higher like you said, but knowing what checks may want to be overridden is difficult to guess.

peppy replied:
Just to keep things simple let's split the base ones out into their own verifier for now, I think. We can deal with exceptions at a later date if required.

Naxesss committed:
Remove unused import

Naxesss committed:
Add reference for screen bounding box numbers

Naxesss committed:
Make `BeatmapVerifier` an interface

Naxesss committed:
Add `CheckBackground` to `OsuBeatmapVerifier`

Naxesss committed:
Rename `BeatmapVerifier` -> `IBeatmapVerifier`

peppy commented on osu.Game/Screens/Edit/Verify/VerifyScreen.cs:
```diff
@@ -0,0 +1,137 @@
+    public class VerifyScreen : EditorScreen
+    {
+        private Ruleset ruleset;
+        private static IBeatmapVerifier beatmapVerifier;
```
as a heads up, `static` is banned in this project for use cases like this.

Naxesss replied:
right, I was intending to ask about this but forgot. looking at how you solved it answers my questions though

peppy committed:
Split out general checks into its own verifier class (and remove `static` usage)

peppy committed:
Fix some styling issues with the verify screen layout

peppy commented:
@Naxesss I've made changes to remove the need to add `General` checks inside the ruleset `BeatmapVerifier`. This was the direction I was intending to propose earlier today.

Only issue remaining I can see is that a lot of the drawable screen logic is a direct copy-paste from the timing screen. I think we can move this out into a common class (at least for the table styling). Will look at doing this later today.

Naxesss commented on osu.Game/Rulesets/Edit/BeatmapVerifier.cs:
```diff
@@ -0,0 +1,24 @@
+
+namespace osu.Game.Rulesets.Edit
+{
+    /// <summary>
+    /// A ruleset-agnostic beatmap converter that identifies issues in common metadata or mapping standards.
+    /// </summary>
+    public class BeatmapVerifier : IBeatmapVerifier
```
"beatmap converter"?

peppy replied:
oops.

peppy committed:
Fix typo in xmldoc

Naxesss committed:
Add missing period

peppy committed:
Merge branch 'master' into verify-tab

peppy committed:
Split out common EditorTable base class

peppy committed:
Move common table layout logic into `EditorTable` abstract class

peppy commented:
I've separated the table layout logic out, and think this looks fine for a first pass.

peppy committed:
Add xmldoc to interfaces

peppy commented:
Let's get this one in. Going forward, if you add more checks please add them in small batches. There's still some missing features which I'd like to see present in the verify system (generally just aiming to improve the user experience and reducing turn-around time in issues being surfaced, hopefully in compose mode) so don't want to get too far before realising we've missed something.
